### PR TITLE
Ensure that errors are located in the current command.

### LINKED ISSUE #15298: Notations involving tactics in terms print incorrect error locations
#### Description of the problem

Consider `foo.v`:
```coq
Notation foo := (I I).

Check foo.
```
Then `coqc -q foo.v` gives
```
File "./foo.v", line 3, characters 6-9:
Error: Illegal application (Non-functional construction):
The expression "I" of type "True" cannot be applied to the term
 "I" : "True"
```
which is correct.

However, consider `bar.v`:
```coq
Notation foo := (ltac:(exact (I I))) (only parsing).

Check foo.
```
Then `coqc -q bar.v` gives
```
File "./bar.v", line 1, characters 30-33:
Error: Illegal application (Non-functional construction):
The expression "I" of type "True" cannot be applied to the term
 "I" : "True"
```
This error location is incorrect.  It does not point to the line where the error occurred, and moreover will point to the file where the notation is defined rather than where it is used.  This makes it impossible for the bug minimizer to automatically minimize files with this sort of error.

#### Coq Version

8.14, also master (57325dc2450aaf51be16f8473f4a4e5ad5d2e7af)

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #15298

Why did we expose Vernacinterp.with_fail?

### CODE REVIEW

File: vernac/vernacinterp.mli [hunk 1]
```diff
@@ -20,8 +20,8 @@ val interp_qed_delayed_proof
   -> Vernacexpr.proof_end CAst.t
   -> Vernacstate.t
 
-(** [with_fail ~st f] runs [f ()] and expects it to fail, otherwise it fails. *)
-val with_fail : st:Vernacstate.t -> (unit -> 'a) -> unit
+(** [with_fail ~loc ~st f] runs [f ()] and expects it to fail, otherwise it fails. *)
+val with_fail : loc:Loc.t option -> st:Vernacstate.t -> (unit -> 'a) -> unit
```
Comment (ppedrot) [inline vernac/vernacinterp.mli:24]: Why not a `?loc` instead?

Comment (SkySkimmer) [reply vernac/vernacinterp.mli:24]: Because we don't want to forget to pass it

**Comment by ejgallego** [comment]
> Why did we expose Vernacinterp.with_fail?

Used to be called from Stm, not anymore :)

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@ejgallego if you're fine with the current state of the PR I can merge now.

**Review by Alizter — APPROVED** [review]

**Comment by ejgallego** [comment]
> @ejgallego if you're fine with the current state of the PR I can merge now.

Looks fine to me, I trust @SkySkimmer to do the right thing here.

**Comment by ejgallego** [comment]
> @ejgallego if you're fine with the current state of the PR I can merge now.

Looks fine to me, I trust @SkySkimmer to do the right thing here.

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**